File: section_two/java/leetcode/src/main/java/com/whg/hard/serializeanddeserializebinarytree/Codec.java
package com.whg.hard.serializeanddeserializebinarytree;

import java.util.ArrayList;
import java.util.List;
/**
 * Solution class
 *
 * problemId 297
 * @author wanghaogang
 * @date 2018/6/18
 */
public class Codec {

    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode(int x) { val = x; }
    }

    private String preStr(TreeNode root) {
        if(root==null) {
            return "*";
        }
        StringBuilder ret = new StringBuilder();
        ret.append(root.val);
        ret.append(",");
        ret.append(preStr(root.left));
        ret.append(",");
        ret.append(preStr(root.right));
        return ret.toString();
    }
    public class Pair {
        TreeNode first;
        int second;
        Pair(TreeNode first,int second) {
            this.first=first;
            this.second=second;
        }
    }
    private Pair parse(String[] data, int cur) {
        String nullChar = "*";
        if(data==null||cur>=data.length||nullChar.equals(data[cur])) {
            return new Pair(null,cur+1);
        }
        TreeNode root = new TreeNode(Integer.parseInt(data[cur]));
        Pair pair = parse(data, cur+1);
        root.left=pair.first;
        pair = parse(data, pair.second);
        root.right=pair.first;
        pair.first=root;
        return pair;
    }
    public String serialize(TreeNode root) {
        return preStr(root);
    }

    public TreeNode deserialize(String data) {
        return parse(data.split(","),0).first;
    }

    public static void main(String[] args) {
        Integer[] arr={3,2,4,3};
        Codec codec = new Codec();
        TreeNode[] nodes = new TreeNode[arr.length];
        for(int i=0;i<arr.length;++i) {
            if(arr[i]!=null) {
                nodes[i]=codec.new TreeNode(arr[i]);
            }
        }
        for(int i=0;i<arr.length;++i) {
            if(arr[i]!=null) {
                if(2*i+1<arr.length&&arr[2*i+1]!=null) {
                    nodes[i].left=nodes[2*i+1];
                }
                if(2*i+2<arr.length&&arr[2*i+2]!=null) {
                    nodes[i].right=nodes[2*i+2];
                }
            }
        }
        System.out.println(codec.serialize(nodes[0]));
        System.out.println(codec.serialize(codec.deserialize(codec.serialize(nodes[0]))));
    }
}

File: section_two/python/leetcode/hard/serializeanddeserializebinarytree/Codec.py
"""
 Solution class

 problemId 297
 @author wanghaogang
 @date 2018/6/29

"""


class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.

        :type root: TreeNode
        :rtype: str
        """
        return self.pre_order_str(root)

    def deserialize(self, data):
        """Decodes your encoded data to tree.

        :type data: str
        :rtype: TreeNode
        """
        return self.parse(data.split(','), 0)[0]

    def pre_order_str(self, root):
        if not root:
            return "*"
        ret = str()
        ret += str(root.val)
        ret += ","
        ret += self.pre_order_str(root.left)
        ret += ","
        ret += self.pre_order_str(root.right)
        return ret

    def parse(self, data, cur):
        if not data or cur >= len(data) or '*' == data[cur]:
            return [None, cur + 1]
        root = TreeNode(int(data[cur]))
        pair = self.parse(data, cur + 1)
        root.left = pair[0]
        pair = self.parse(data, pair[1])
        root.right = pair[0]
        pair[0] = root
        return pair
